Change showcase background color based on scroll position

joeowenca opened the pull request:
### Files added
Added a calculateScrollAlpha.ts file to contain the script that calculates the alpha value based on the scroll position.

### Things changed
The showcase component has been updated to import the alpha value from calculateScrollAlpha.ts. Then, using useEffect and event listeners, the background color of the showcase component is updated based on the alpha value and scroll position.

In order for the showcase background to take up the entire page, padding is used on the element. However, this blocks elements behind it. Therefore, some additional code is added so that when the alpha value is set to "0.00" (aka, showcase is hidden), the padding on the element is swapped out for margin so that showcase retains its position, but elements below it can be selected and highlighted.

joeowenca committed:
Add script to calculate alpha based on scroll position

joeowenca committed:
Update showcase background color based on scroll position

joeowenca committed:
Update showcase item padding

joeowenca committed:
Set return type of calculateScrollAlpha to string

joeowenca committed:
Hide showcase when alpha value is 0.00

To make the showcase background color affect the whole page, padding is
used. However, since the padding spans the entire height of the page,
the introduction component can't be selected or highlighted. When the
alpha value is "0.00", padding is swapped out for margin to ensure that
elements below the showcase component can be selected while it's hidden.

brettgoss commented on src/app/page.tsx:
```diff
@@ -1,3 +1,5 @@
+'use client';
```
Never seen this before, what does it do?

joeowenca replied:
Haha so this was a question I was gonna ask you. but reading up on the Next.js [rendering doc](https://nextjs.org/docs/app/building-your-application/rendering), Next.js is opt-in to client-rendered components. Therefore, all components by default are rendered on the server and thus can't handle user state (setting state based on scroll position) like a component rendered on the client-side would. 

I'm just specifying `use client` on the parent component so all child components linked to it will be rendered client-side.

brettgoss replied:
Ahh okay cool. I actually have needed to use clientside rendering in my Next.js project before, but I just used `next/dynamic` with the `ssr: false` [option](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading#with-no-ssr). But this is definitely cleaner!

Though it might be worth moving this down to just where you need it so that you can take advantage of the [other benefits of server-side rendering](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading#with-no-ssr)

brettgoss commented on src/app/components/showcase.tsx:
```diff
@@ -1,6 +1,37 @@
+	function handleScroll() {
+		let scrollAlpha: string = calculateScrollAlpha();
+
+		if (scrollAlpha === '0.00') {
+			setHidden(true);
+		} else {
+			setHidden(false);
+		}
```
Could this be re-written as just `setHidden(scrollAlpha === '0.00');`?

joeowenca replied:
Yup - this is why I'm thankful for your code reviews haha. In the moment I couldn't think of writing this a better way, and I haven't used your suggested syntax enough (will now). 

But that makes perfect sense now to just set state to whatever the condition `scrollAlpha === '0.00'` outputs to (`true` or `false`). 

Cleaned up in a recent commit!

brettgoss commented on src/app/components/showcase.tsx:
```diff
@@ -1,6 +1,37 @@
+	const [hidden, setHidden] = useState(true);
+
+	function handleScroll() {
+		let scrollAlpha: string = calculateScrollAlpha();
```
I think this could probably be a const since it's not being reassigned? Also probably no need to cast it to a `: string` since the function itself is typed to return a string?

joeowenca replied:
Ahh so I was completely thinking about `let` and `const` in the wrong way in this context. I had originally created `scrollAlpha` as `let` since I thought it was getting continuously reassigned by `calculateScrollAlpha()`. But it's only getting its value assigned once per function call, so I see now how `const` works better in this case. Thanks for pointing that out!

So, maybe just beginner things with TypeScript, but in this case I was just trying to type everything I could. So, if I'm assigning a variable a set value, like `const text = "string of text"`, I don't need to cast it to a type? 

Or, should that be typed, and I only don't need to type `let scrollAlpha: string = calculateScrollAlpha();` since `calculateScrollAlpha()` already has a return type specified of string?

brettgoss replied:
For let vs const: yeah exactly, the `const` is scoped to the function itself, and it's never changing within that function call. You could even probably just get rid of your `hidden` state entirely also, and just check `alpha === '0.00'` where you need it. 

As for the typing, you're exactly right, the cast is unneeded since the function itself is already saying that it returns a string. Generally it's better to just let Typescript infer the types as long as it's able to do so correctly. So if your IDE can tell that the return value is a string, then you shouldn't need to explicitly declare that the variable is a string. But if typescript is saying the variable is `any` or `unknown`, then it's a sign you're missing a type declaration somewhere.

You can even enable a rule in your typescript config for `noImplicitAny` which will warn/error if there's anything with an ambiguous type. It can be a bit annoying at times because it's pretty strict, but it allows you to spot areas where the types are not well defined. 

This article goes into more depth about type inference with typescript: https://davidgomes.com/annotate-vs-type-inference/

brettgoss commented on src/scripts/calculateScrollAlpha.ts:
```diff
@@ -0,0 +1,6 @@
+export default function calculateScrollAlpha():string {
```
I think you could probably have this return `: number` since `.toFixed()` returns a float, which is a number in TS. It probably will still work as a string, but using the numeric value is probably easier and more precise. Eg `0.00` instead of `'0.00'`

joeowenca replied:
I initially thought that it was float too (given two decimal places), but checking the [docs for toFixed()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed#return_value) I saw its return type is string. 

I think I'd then need to parse it to a float, so instead of adding that additional code, and since I'm not doing any further calculations on the alpha value, I kept my code as having alpha as a string - which seems to work in my application of the alpha value. 

Do you think it would be beneficial to change `alpha` to a float instead of string?

brettgoss replied:
Floats are sort of halfway between a string and an int, so generally they can go either way. You're right about `.toFixed()` though, my bad, it does seem to return a true string. 

The biggest difference is precision, and how easy they are to work with for calculations. Consider the following examples:
```js
10.00 === 10 // true
10.00 === 10.00 // true
10.00 + 12.00 // 22
'10.00' + '12.00' // '10.0012.00'
'10.00' === 10 // false
'10.00' === '10.0' // false
'10.00' === '10.00' // true
```
So as a string, there's a difference between `'10.0'` and `'10.00'` which is useful if you want to display the decimal points (eg. when displaying money), however as a float/number, `10`, `10.0` and `10.000000` are all equivalent, which is useful if you're doing math or calculations. 

All that to say, it's totally your call depending on the application, but unless you're displaying the value, I generally stick with the more numeric option.

joeowenca committed:
Clean up handleScroll function in showcase.tsx

joeowenca committed:
Change return type of calculateScrollAlpha from string to number

Having calculateScrollAlpha return a number instead of a string will
help perform further calculations on its value.

joeowenca committed:
Remove hidden state and check the condition on the className directly

joeowenca committed:
Move 'use client' down from parent to only the children that need it

joeowenca committed:
Fix issue of window being referenced on server component
